### PR TITLE
Fix some circular import/incremental mode bugs

### PR DESCRIPTION
**Pull request by Michael0x2a**
This is a work-in-progress, but I just wanted to make sure any small bugfixes I might make are vetted periodically/make sure that the new tests continue to pass CI.

### CODE REVIEW

**Comment by Michael0x2a** [comment]
This pull request fixes two bugs related to how import cycles and submodules are handled by mypy during incremental mode.

The first bug doesn't have an issue associated with it, the second fixes the bug in issue https://github.com/python/mypy/issues/1774.

**Comment by gvanrossum** [comment]
Good work! I adjusted the commit message a bit.

File: mypy/test/testcheck.py [hunk 1]
```diff
@@ -187,8 +187,11 @@ def find_module_files(self) -> Dict[str, str]:
             del dnparts[0]
             for file in files:
                 if file.endswith('.py'):
-                    base, ext = os.path.splitext(file)
-                    id = '.'.join(dnparts + [base])
+                    if file == "__init__.py":
```
Comment (ddfisher) [inline mypy/test/testcheck.py]: Why are we special-casing `__init__` files here?  Could you add a comment?

Comment (Michael0x2a) [inline mypy/test/testcheck.py]: @ddfisher -- I submitted a pull request with the comment here: https://github.com/python/mypy/pull/1844

Comment (ddfisher) [inline mypy/test/testcheck.py]: Great, thanks!

**Pull request merged**